### PR TITLE
Rewrite dropck to be more correct

### PR DESCRIPTION
**Pull request by arielb1**
This fixes multiple bugs, and as several of these are soundness issue, is a [breaking-change].

r? @pnkfelix

### CODE REVIEW

**Comment by pnkfelix** [comment]
I have no problem with the refactoring overall.  E.g. I have never been wedded to the use of the `ty::walk` method, so I do not mind the use of an explicit `match` with each case handled explicitly.

I am concerned about the removal of the handling of `ReStatic`.  Maybe I overlooked something in your approach, but that side-condition was definitely motivated by concrete cases I encountered -- I would be a little surprised if this PR is actually passing the `make check` test suite, but maybe we never put in regression tests for the issue in question.  :(

**Comment by arielb1** [comment]
My main opposition to the `ReStatic` check is that I don't think it is needed - if you have `Trait+'static`, then it being a dtorck type would only mean that we check `Trait+'static : 'destruction_scope`, which is totally harmless - in fact, in the case of `Trait+'tcx`, we already do the check.

With the current state of `type-outlives`, the check should trivially hold. However, with the new `type-outlives`, this would also forbid types like `Trait<B<'a>>+'static`, which are an important component of #26656.

**Comment by arielb1** [comment]
[rebased]

File: src/librustc_typeck/check/dropck.rs [hunk 1]
```diff
@@ -324,269 +328,200 @@ enum TypeContext {
...
     }
 
-    let origin = || infer::SubregionOrigin::SafeDestructor(span);
-    let mut walker = ty_root.walk();
-    let opt_phantom_data_def_id = rcx.tcx().lang_items.phantom_data();
+    let opt_phantom_data_def_id = tcx.lang_items.phantom_data();
 
-    let destructor_for_type = rcx.tcx().destructor_for_type.borrow();
+    // FIXME(arielb1): don't be O(n^2)
+    if cx.breadcrumbs.contains(&ty) {
```
Comment (eddyb) [inline src/librustc_typeck/check/dropck.rs]: Why not use `FnvHashSet<Ty<'tcx>>`?

**Comment by pnkfelix** [comment]
(okay, I'm now curious why the heck I added `ReStatic` in the first place, but I cannot argue with the logic you have presented here.)

**Comment by pnkfelix** [comment]
(while this is a `[breaking-change]`, I think the problems it fixes are of a nature that we can safely just land this without concerning ourselves with e.g. warning cycles et cetera)

**Comment by brson** [comment]
Marking this as stable-regression so i can find it later, even though these are just soundness fixes.

**Pull request merged**